### PR TITLE
adds ability to specify multiple regex strings for a single fact

### PR DESCRIPTION
**Pull request by logicminds**
- previously if you wanted to match multiple hosts across a large network
  your regex pattern has to be extremely crafty.  With this feature
  you can now supply multiple regex values to match against, thus making your regex
  patterns more trivial.

### CODE REVIEW

**Comment by logicminds** [comment]
I wrote the tests for this but I have no idea how to run the tests.  There is no rakefile or docs on how to run run your tests.  I can document this once I know.

**Comment by GregSutcliffe** [comment]
Hey, thanks for this!

For running the tests, see http://projects.theforeman.org/projects/foreman/wiki/How_to_Create_a_Plugin#Testing - in short, add the plugin to a normal Foreman source install and run "rake test" or "rake test:default_hostgroup"

I'll add some comments inline, but overall it looks great.

File: .gitignore [hunk 1]
```diff
@@ -1,3 +1,5 @@
 # Need a config/environment.rb for rails.vim to work in engines
 config/environment.rb
 *gem
+.idea
+.bundle
```
Comment (GregSutcliffe) [inline .gitignore]: Seems like an unnecessary change? There's no reason to run bundler from within a Foreman plugin.

Comment (logicminds) [inline .gitignore]: Out of habit I always had .bundle since the rest of my repos need it.  I can remove it.

File: default_hostgroup.yaml.example [hunk 2]
```diff
@@ -18,6 +18,8 @@
 ---
 :default_hostgroup:
   :facts_map:
+    "Go_agents":
+      "hostname": ["go_agent1", "go_agent2", "go_agent[1][1-3]]"
```
Comment (GregSutcliffe) [inline default_hostgroup.yaml.example]: This is not the ideal way to specify an array in YAML (although it does work). Better to use

```
"Go_agents":
- "go_agent1"
- "go_agent2"
- "go_agent[1][1-3]"
```

Comment (logicminds) [inline default_hostgroup.yaml.example]:
```
"Go_agents":
  "hostname":
    - "go_agent1"
    - "go_agent2"
    - "go_agent[1][1-3]"
```

Did you mean this way?

File: lib/default_hostgroup_managed_host_patch.rb [hunk 3]
```diff
@@ -50,11 +50,14 @@ def import_host_and_facts_with_match_hostgroup hostname, facts, certname = nil,
 
     def group_matches?(fact)
       fact.each do |fact_name, fact_regex|
-        fact_regex.gsub!(/(\A\/|\/\z)/, '')
+        fact_regexs = Array(fact_regex)
         host_fact_value = @host.facts_hash[fact_name]
-        Rails.logger.info "Fact = #{fact_name}"
-        Rails.logger.info "Regex = #{fact_regex}"
-        return true if Regexp.new(fact_regex).match(host_fact_value)
+        fact_regexs.each do |fact_regex|
```
Comment (GregSutcliffe) [inline lib/default_hostgroup_managed_host_patch.rb]: You could combine this with line 53 as `[fact_regex].each do |regex|`

File: test/unit/default_hostgroup_test.rb [hunk 4]
```diff
@@ -24,6 +24,28 @@ def setup_hostgroup_match
...
+    SETTINGS[:default_hostgroup][:facts_map] = { "Go Agents" => {'hostname' => ["/go_agent1/", "/sinn1636.lan/", "/go_agent[1][1-3]/"] }}
+
+    hostgroup = Hostgroup.create(:name => "Go Agents")
+    raw = parse_json_fixture('/facts.json')
+
+    assert Host.import_host_and_facts(raw['name'], raw['facts'])
+    assert_equal hostgroup, Host.find_by_name('sinn1636.lan').hostgroup
+  end
+
+  test "multiple regex matches for same fact enclosed in /" do
```
Comment (GregSutcliffe) [inline test/unit/default_hostgroup_test.rb]: This has the same test name as line 27, i think you mean `not enclosed in /`

File: README.md [hunk 5]
```diff
@@ -35,15 +35,31 @@ your needs. The format is shown in the example. The simplest form would be:
 ```
 
 *Important Note:* You have to restart foreman in order to apply changes in
-`default_hostgroup.yaml`!
+`foreman_default_hostgroup.yaml`!
```
Comment (GregSutcliffe) [inline README.md]: I don't see a rename of `default_hostgroup.yaml` in this PR, either drop this change or `git mv` the file :)

File: default_hostgroup.yaml.example (hunk 2 above)
Comment (GregSutcliffe) [inline default_hostgroup.yaml.example]: Sorry, yes. Rushed that comment slightly.

File: .gitignore (hunk 1 above)
Comment (GregSutcliffe) [inline .gitignore]: I'm not that bothered, it's a nitpick, but it feels like commit churn. If it was the only thing I'd pass, but since you're going to be squashing the other things... ;)

File: README.md (hunk 5 above)
Comment (GregSutcliffe) [inline README.md]: Actually, furthermore, you'll need to update the gemspec and possible the rpm/deb packaging too - is there a good reason behind this rename?

Comment (logicminds) [inline README.md]: I renamed it because the docs where out of date.  The filename is actually foreman_default_hostgroup and that is how the gem/rpm installs it.  Having the incorrect name here will confuse people and the plugin will not even function (afaik).

Since I was updating the docs I thought I would just fix a few other things that needed updating.

File: lib/default_hostgroup_managed_host_patch.rb (hunk 3 above)
Comment (logicminds) [inline lib/default_hostgroup_managed_host_patch.rb]: I still need to use Array(), otherwise I would have to call flatten since a nested array will occur whereas Array() takes care of this for me automatically.

File: README.md (hunk 5 above)
Comment (logicminds) [inline README.md]: @GregSutcliffe is this correct?  The rpm installs this as foreman_default_hostgroup.yaml but that must be just the rpm.  So I am not sure what to do here.  I can update the gem spec if needed and change the filename.

**Comment by logicminds** [comment]
@GregSutcliffe I am having trouble testing.  Is the load path not setup correctly?  This occurs when running rake test:default_hostgroup.

``` shell
/opt/rh/ruby193/root/usr/local/share/gems/bundler/gems/foreman_default_hostgroup-044ae9fec6a7/test/unit/default_hostgroup_test.rb:1:in `require': cannot load such file -- test_plugin_helper (LoadError)
        from /opt/rh/ruby193/root/usr/local/share/gems/bundler/gems/foreman_default_hostgroup-044ae9fec6a7/test/unit/default_hostgroup_test.rb:1:in `<top (required)>'
        from /opt/rh/ruby193/root/usr/local/share/gems/gems/rake-10.4.2/lib/rake/rake_test_loader.rb:10:in `require'
        from /opt/rh/ruby193/root/usr/local/share/gems/gems/rake-10.4.2/lib/rake/rake_test_loader.rb:10:in `block (2 levels) in <main>'
        from /opt/rh/ruby193/root/usr/local/share/gems/gems/rake-10.4.2/lib/rake/rake_test_loader.rb:9:in `each'
        from /opt/rh/ruby193/root/usr/local/share/gems/gems/rake-10.4.2/lib/rake/rake_test_loader.rb:9:in `block in <main>'
        from /opt/rh/ruby193/root/usr/local/share/gems/gems/rake-10.4.2/lib/rake/rake_test_loader.rb:4:in `select'
        from /opt/rh/ruby193/root/usr/local/share/gems/gems/rake-10.4.2/lib/rake/rake_test_loader.rb:4:in `<main>'


```

File: README.md (hunk 5 above)
Comment (logicminds) [inline README.md]: I decided to just drop this change and have updated the commit to not include these changes.

**Comment by GregSutcliffe** [comment]
@logicminds ugh sorry this slipped my mind - thanks for reminding me. PR looks great now - I'll try to test over the weekend. Are you still having test errors yourself?

**Comment by logicminds** [comment]
Yes, I am still having test errors.  I have another PR that tries to automate the testing with travis much like foreman hooks.  Give me a minute to push that.

**Comment by logicminds** [comment]
See #15 regarding testing.

**Comment by GregSutcliffe** [comment]
Commented on #15. Regarding the above failure, which dir are you in when running the tests, and what's the full command you're using?

EDIT: also, is the test_plugin_helper file present in the test dir of the plugin? (I assume so but lets check those assumptions :P)

**Comment by logicminds** [comment]
Running the rake test command and the rake test default host group .  It could be some dev environment setting I don't have that you might.

**Comment by GregSutcliffe** [comment]
> Running the rake test command and the rake test default host group . It could be some dev environment setting I don't have that you might.

And that's from your Foreman dir, rather than from the plugin dir, right? And the file is present?

**Comment by logicminds** [comment]
That's from the foreman directory

**Comment by ares** [comment]
I appologize, this plugin has not been maintained for a long time. Given it has now conflicts, I'm closing. I'll try to find an active maintainer, please consider reopening after the rebase.
